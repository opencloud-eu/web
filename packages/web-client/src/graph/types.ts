import type { AxiosInstance, ResponseType } from 'axios'
import type { Configuration } from './generated'

export interface GraphFactoryOptions {
  axiosClient: AxiosInstance
  config: Configuration
}

export interface GraphRequestOptions {
  headers?: Record<string, string>
  params?: Record<string, string>
  signal?: AbortSignal
  responseType?: ResponseType
}
