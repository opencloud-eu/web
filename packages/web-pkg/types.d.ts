import { Ability } from '@opencloud-eu/web-client'
import {
  ArchiverService,
  ClientService,
  LoadingService,
  PreviewService,
  PasswordPolicyService
} from './src/services'
import { Route, Router } from 'vue-router'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
    $passwordPolicyService: PasswordPolicyService
    $router: Router
    $route: Route
  }
}
