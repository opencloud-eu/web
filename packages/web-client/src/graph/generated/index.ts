/* tslint:disable */
/* eslint-disable */
/**
 * Libre Graph API
 * Libre Graph is a free API for cloud collaboration inspired by the MS Graph API.
 *
 * The version of the OpenAPI document: v1.0.7
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

