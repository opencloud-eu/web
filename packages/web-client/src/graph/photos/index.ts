export * from './photo'
export * from './types'
