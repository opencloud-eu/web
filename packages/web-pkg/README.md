# web-pkg

`web-pkg` is a package that provides utilities, most importantly a variety of components and composables, that can be useful when developing apps and extensions for OpenCloud Web.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-pkg --save-dev

$ pnpm add -D @opencloud-eu/web-pkg

$ yarn add @opencloud-eu/web-pkg --dev
```

It's recommended to install this package as a dev dependency because it's only really needed for providing autocompletion in your IDE and unit tests. In a runtime context, the OpenCloud Web runtime provides the actual implementation.
