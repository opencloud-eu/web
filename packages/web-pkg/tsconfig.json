{
  "extends": "@opencloud-eu/tsconfig",

  // necessary for building web-pkg standalone, otherwise we get errors with inferred types
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@vue/shared": ["node_modules/@vue/shared/"],
      "@codemirror/autocomplete": ["./node_modules/@codemirror/autocomplete"],
      "vue/jsx-runtime": ["./node_modules/vue/jsx-runtime"],
      "@popperjs/core": ["./../design-system/node_modules/@popperjs/core"]
    }
  },

  // include component types of the design-system
  "include": ["../design-system/types.d.ts"]
}
