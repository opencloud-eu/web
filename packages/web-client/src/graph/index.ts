import { AxiosInstance } from 'axios'
import { Configuration } from './generated'
import { type GraphUsers, UsersFactory } from './users'
import { type GraphGroups, GroupsFactory } from './groups'
import { ApplicationsFactory, GraphApplications } from './applications'
import { DrivesFactory, GraphDrives } from './drives'
import { DriveItemsFactory, GraphDriveItems } from './driveItems'
import { TagsFactory, GraphTags } from './tags'
import { ActivitiesFactory, GraphActivities } from './activities'
import { PermissionsFactory, GraphPermissions } from './permissions'
import { PhotoFactory, GraphPhotos } from './photos'

export interface Graph {
  activities: GraphActivities
  applications: GraphApplications
  tags: GraphTags
  drives: GraphDrives
  driveItems: GraphDriveItems
  users: GraphUsers
  groups: GraphGroups
  permissions: GraphPermissions
  photos: GraphPhotos
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  return <Graph>{
    activities: ActivitiesFactory({ axiosClient, config }),
    applications: ApplicationsFactory({ axiosClient, config }),
    tags: TagsFactory({ axiosClient, config }),
    drives: DrivesFactory({ axiosClient, config }),
    driveItems: DriveItemsFactory({ axiosClient, config }),
    users: UsersFactory({ axiosClient, config }),
    groups: GroupsFactory({ axiosClient, config }),
    permissions: PermissionsFactory({ axiosClient, config }),
    photos: PhotoFactory({ axiosClient, config })
  }
}
