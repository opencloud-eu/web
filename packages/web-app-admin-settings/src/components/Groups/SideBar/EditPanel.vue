<template>
  <div id="group-edit-panel" class="mt-12">
    <group-info-box :group="group" />
    <form
      id="group-edit-form"
      class="bg-role-surface-container p-4 rounded-t-sm"
      autocomplete="off"
    >
      <oc-text-input
        id="displayName-input"
        v-model="editGroup.displayName"
        class="mb-2"
        :label="$gettext('Group name')"
        :error-message="formData.displayName.errorMessage"
        :fix-message-line="true"
        required-mark
        @update:model-value="validateDisplayName"
      />
      <compare-save-dialog
        class="mb-6 rounded-b-sm"
        :original-object="group"
        :compare-object="editGroup"
        :confirm-button-disabled="invalidFormData"
        @revert="revertChanges"
        @confirm="onEditGroup(editGroup)"
      ></compare-save-dialog>
    </form>
  </div>
</template>
<script lang="ts">
import { defineComponent, PropType, ref } from 'vue'
import { Group } from '@opencloud-eu/web-client/graph/generated'
import { CompareSaveDialog, eventBus, useMessages } from '@opencloud-eu/web-pkg'
import { MaybeRef, useClientService } from '@opencloud-eu/web-pkg'
import GroupInfoBox from './GroupInfoBox.vue'
import { useGroupSettingsStore } from '../../../composables'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'EditPanel',
  components: {
    GroupInfoBox,
    CompareSaveDialog
  },
  props: {
    group: {
      type: Object as PropType<Group>,
      required: true,
      default: null
    }
  },
  emits: ['confirm'],
  setup() {
    const clientService = useClientService()
    const { showErrorMessage } = useMessages()
    const groupSettingsStore = useGroupSettingsStore()
    const { $gettext } = useGettext()

    const editGroup: MaybeRef<Group> = ref({})
    const formData = ref({
      displayName: {
        errorMessage: '',
        valid: true
      }
    })

    const onEditGroup = async (editGroup: Group) => {
      try {
        const client = clientService.graphAuthenticated
        await client.groups.editGroup(editGroup.id, editGroup)
        const updatedGroup = await client.groups.getGroup(editGroup.id)
        groupSettingsStore.upsertGroup(updatedGroup)

        eventBus.publish('sidebar.entity.saved')

        return updatedGroup
      } catch (error) {
        console.error(error)
        showErrorMessage({
          title: $gettext('Failed to edit group'),
          errors: [error]
        })
      }
    }

    return {
      clientService,
      editGroup,
      formData,
      onEditGroup
    }
  },
  computed: {
    invalidFormData() {
      return Object.values(this.formData)
        .map((v: any) => !!v.valid)
        .includes(false)
    }
  },
  watch: {
    group: {
      handler: function () {
        this.editGroup = { ...this.group }
      },
      deep: true,
      immediate: true
    }
  },
  methods: {
    async validateDisplayName() {
      this.formData.displayName.valid = false

      if (this.editGroup.displayName.trim() === '') {
        this.formData.displayName.errorMessage = this.$gettext('Group name cannot be empty')
        return false
      }

      if (this.editGroup.displayName.length > 255) {
        this.formData.displayName.errorMessage = this.$gettext(
          'Group name cannot exceed 255 characters'
        )
        return false
      }

      if (this.group.displayName !== this.editGroup.displayName) {
        try {
          const client = this.clientService.graphAuthenticated
          await client.groups.getGroup(this.editGroup.displayName)
          this.formData.displayName.errorMessage = this.$gettext(
            'Group "%{groupName}" already exists',
            {
              groupName: this.editGroup.displayName
            }
          )
          return false
        } catch {}
      }

      this.formData.displayName.errorMessage = ''
      this.formData.displayName.valid = true
      return true
    },
    revertChanges() {
      this.editGroup = { ...this.group }
      Object.values(this.formData).forEach((formDataValue: any) => {
        formDataValue.valid = true
        formDataValue.errorMessage = ''
      })
    }
  }
})
</script>
