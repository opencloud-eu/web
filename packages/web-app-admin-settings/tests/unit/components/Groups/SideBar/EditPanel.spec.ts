import EditPanel from '../../../../../src/components/Groups/SideBar/EditPanel.vue'
import {
  defaultComponentMocks,
  defaultPlugins,
  mockAxiosReject,
  mount
} from '@opencloud-eu/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { eventBus, useMessages } from '@opencloud-eu/web-pkg'
import { Group } from '@opencloud-eu/web-client/graph/generated'

describe('EditPanel', () => {
  it('renders all available inputs', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  describe('method "revertChanges"', () => {
    it('should revert changes on property editGroup', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editGroup.displayName = 'users'
      wrapper.vm.revertChanges()
      expect(wrapper.vm.editGroup.displayName).toEqual('group')
    })
    it('should revert changes on property formData', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = false
      wrapper.vm.formData.displayName.errorMessage = 'error'
      wrapper.vm.revertChanges()
      expect(wrapper.vm.formData.displayName.valid).toBeTruthy()
      expect(wrapper.vm.formData.displayName.errorMessage).toEqual('')
    })
  })

  describe('method "validateDisplayName"', () => {
    it('should return true if displayName is valid', async () => {
      const { wrapper, mocks } = getWrapper()
      wrapper.vm.editGroup.displayName = 'users'
      const graphMock = mocks.$clientService.graphAuthenticated
      const getGroupStub = graphMock.groups.getGroup.mockRejectedValue(() => mockAxiosReject())
      expect(await wrapper.vm.validateDisplayName()).toBeTruthy()
      expect(getGroupStub).toHaveBeenCalled()
    })
    it('should return false if displayName is longer than 255 characters', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editGroup.displayName = 'n'.repeat(256)
      expect(await wrapper.vm.validateDisplayName()).toBeFalsy()
    })
    it('should return false if displayName is empty', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editGroup.displayName = ''
      expect(await wrapper.vm.validateDisplayName()).toBeFalsy()
    })
    it('should return false if displayName is already existing', async () => {
      const { wrapper, mocks } = getWrapper()
      wrapper.vm.editGroup.displayName = 'users'
      const graphMock = mocks.$clientService.graphAuthenticated
      const getGroupStub = graphMock.groups.getGroup.mockResolvedValue(
        mock<Group>({ displayName: 'group' })
      )
      expect(await wrapper.vm.validateDisplayName()).toBeFalsy()
      expect(getGroupStub).toHaveBeenCalled()
    })
  })

  describe('method "onEditGroup"', () => {
    it('should emit event on success', async () => {
      const { wrapper, mocks } = getWrapper()

      const clientService = mocks.$clientService
      clientService.graphAuthenticated.groups.editGroup.mockResolvedValue()
      clientService.graphAuthenticated.groups.getGroup.mockResolvedValue(
        mock<Group>({ id: '1', displayName: 'administrators' })
      )

      const editGroup = {
        id: '1',
        name: 'administrators'
      }

      const busStub = vi.spyOn(eventBus, 'publish')
      const updatedGroup = await wrapper.vm.onEditGroup(editGroup)

      expect(updatedGroup.id).toEqual('1')
      expect(updatedGroup.displayName).toEqual('administrators')
      expect(busStub).toHaveBeenCalled()
    })

    it('should show message on error', async () => {
      vi.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper, mocks } = getWrapper()
      const clientService = mocks.$clientService
      clientService.graphAuthenticated.groups.editGroup.mockRejectedValue(undefined)
      await wrapper.vm.onEditGroup({})

      const { showErrorMessage } = useMessages()
      expect(showErrorMessage).toHaveBeenCalled()
    })
  })

  describe('computed method "invalidFormData"', () => {
    it('should be false if formData is invalid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = true
      expect(wrapper.vm.invalidFormData).toBeFalsy()
    })
    it('should be true if formData is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = false
      expect(wrapper.vm.invalidFormData).toBeTruthy()
    })
  })
})

function getWrapper() {
  const mocks = defaultComponentMocks()

  return {
    mocks,
    wrapper: mount(EditPanel, {
      props: {
        group: { displayName: 'group', members: [] }
      },
      global: {
        mocks,
        provide: mocks,
        plugins: [...defaultPlugins()],
        stubs: {
          'oc-text-input': true,
          'avatar-image': true,
          'oc-button': true,
          translate: true
        }
      }
    })
  }
}
