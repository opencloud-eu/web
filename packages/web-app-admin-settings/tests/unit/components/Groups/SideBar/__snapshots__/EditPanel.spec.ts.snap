// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditPanel > renders all available inputs 1`] = `
"<div id="group-edit-panel" class="oc-mt-xl">
  <div class="oc-flex group-info oc-mb-l"><span class="vue-avatar--wrapper oc-avatar oc-mb-m" style="width: 80px; height: 80px; line-height: 80px; background-color: var(--oc-role-secondary); font-size: 32px; font-family: Helvetica, Arial, sans-serif; color: white;" width="80" aria-hidden="true" focusable="false" data-test-user-name="group"><span class="avatar-initials" style="color: white;">G</span></span> <span class="oc-text-muted group-info-display-name">group</span> <span class="oc-text-muted">0 members</span></div>
  <form id="group-edit-form" class="oc-surface-container oc-p-m" autocomplete="off">
    <oc-text-input-stub id="displayName-input" modelvalue="group" class="oc-mb-s" label="Group name" error-message="" fix-message-line="true" required-mark=""></oc-text-input-stub>
    <div data-v-79550c21="" class="compare-save-dialog oc-width-1-1 oc-flex oc-flex-between oc-flex-middle edit-compare-save-dialog oc-mb-l"><span data-v-79550c21="" class="state-indicator">No changes</span>
      <div data-v-79550c21="">
        <oc-button-stub data-v-79550c21="" disabled="true" class="compare-save-dialog-revert-btn"></oc-button-stub>
        <oc-button-stub data-v-79550c21="" appearance="filled" class="compare-save-dialog-confirm-btn" disabled="true"></oc-button-stub>
      </div>
    </div>
  </form>
</div>"
`;
